# Generated by Django 2.0.7 on 2019-02-28 12:20
import json
import os

from django.db import migrations, connection

DATA_PATH = os.path.join(os.path.dirname(__file__), 'data', '0013')


def forwards_func(apps, schema_editor):
    federal_districts(apps, schema_editor, forward=True)
    federal_regions(apps, schema_editor, forward=True)


def reverse_func(apps, schema_editor):
    federal_regions(apps, schema_editor, forward=False)
    federal_districts(apps, schema_editor, forward=False)


class Migration(migrations.Migration):

    dependencies = [
        ('regions', '0012_auto_20190228_1520'),
    ]

    operations = [
        migrations.RunPython(forwards_func, reverse_func),
    ]


def federal_districts(apps, schema_editor, forward=True):
    FederalDistrict = apps.get_model("regions", "FederalDistrict")
    if not forward:
        FederalDistrict.objects.all().delete()
    else:
        with open(os.path.join(DATA_PATH, 'districts.json')) as f:
            districts = json.load(f)
            for district in districts:
                FederalDistrict.objects.create(**district)

    reset_sequence(FederalDistrict)


def federal_regions(apps, schema_editor, forward=True):
    FederalSubject = apps.get_model("regions", "FederalSubject")
    if not forward:
        FederalSubject.objects.all().update(district=None)
    else:
        with open(os.path.join(DATA_PATH, 'districts_regions.json')) as f:
            districts_regions = json.load(f)
            for district, regions in districts_regions.items():
                FederalSubject.objects.filter(id__in=regions).update(district_id=district)


def reset_sequence(model):
    from django.core.management.color import no_style
    try:
        connection.cursor().execute(connection.ops.sequence_reset_sql(no_style(), [model])[0])
    except IndexError:
        # Only postgresql and oracle backends implement the sequence_reset_sql operation.
        # Other backends (sqlite, mysql) don't have sequences, thus nothing to reset here.
        # So it's safe to swallow this exception here.
        pass
